app/isotope_storage.py

```diff
@@ -0,0 +1,332 @@
+
+    def _path(self, instrument_settings, adduct, charge):
+        dir_path = self._dir_path(instrument_settings, charge)
+        full_path = dir_path / "adduct_{}.parquet".format(adduct)
+        return full_path
+
+    def _load_single_file(self, filename):
+        return pyarrow.parquet.read_table(str(filename)).to_pandas()
+
+    # Caches Pandas dataframes for the last few instrument/polarity configurations.
+    @functools.lru_cache(maxsize=ISOTOPE_CACHE_SIZE)
+    def _load_dir(self, directory):
+        # although pyarrow.parquet theoretically can read multiple files at once,
+        # same Pandas indices seem to confuse it - some column values end up duplicated
+        p = pathlib.Path(directory)
+        dfs = [self._load_single_file(path) for path in p.iterdir()]
+        return pd.concat(dfs)
+
+    def _load(self, filename):
+        p = pathlib.Path(filename)
+        if p.is_dir():  # only one level is supported
+            return self._load_dir(str(p))
+        else:
+            return self._load_single_file(filename)
+
+    def _molecular_formulas(self, db_id=None):
+        """
+        db_id=None means all databases
+        """
+        if db_id is None:
+            db_id = -1
+
+        return self._mf_cache[int(db_id)]
+
+    def _configurations(self):
+        """
+        Represents directory tree as a dataframe with pts_per_mz, charge, adduct columns
+        """
+        result = []
+        for instr_dir in self._dir.iterdir():
+            pts_per_mz = int(instr_dir.name.split('_')[1])
+            for charge_dir in instr_dir.iterdir():
+                charge = int(charge_dir.name.split('_')[1])
+                for adduct_dir in charge_dir.iterdir():
+                    adduct = adduct_dir.name.split('_')[1].split('.')[0]
+                    result.append(dict(pts_per_mz=pts_per_mz, charge=charge, adduct=adduct))
+
+        return pd.DataFrame.from_records(result)
+
+    def adduct_charge_pairs(self):
+        """
+        Set of all (adduct, charge) combinations added so far
+        """
+        return {(x[1], x[2]) for x in self._configurations()[['adduct', 'charge']].itertuples()}
+
+    def instrument_settings(self):
+        """
+        Set of all instrument settings added so far
+        """
+        return {InstrumentSettings(pts_per_mz=x) for x in self._configurations()['pts_per_mz'].unique()}
+
+    def load_patterns(self, instrument_settings, charge, db_id=None, adducts=None):
+        dir_path = self._dir_path(instrument_settings, charge)
+        df = self._load(dir_path)
+        if adducts:
+            adducts_df = pd.DataFrame({'adduct': list(set(adducts))})
+            df = pd.merge(df, adducts_df)  # faster than .isin() method
+        if not db_id:  # = all databases
+            return df
+        mol_formulas = self._molecular_formulas(db_id)
+        return pd.merge(df, mol_formulas)
+
+    def load_fdr_subsample(self, instrument_settings, charge, db_id,
```

Comment:
It feels strange to see fdr terminology here as moldb is meant to operate with molecular databases and isotopic patterns.

Reply:
Feel free to come up with a better name, but this is the best place to perform this subsampling (no need to transfer the whole table over network, wasting time on serialization)